### PR TITLE
TEL comments on day 4 homework

### PR DESCRIPTION
**Pull request by tomlatham**
You gave it a good go but, as you said in your email, it wasn't complete.
Hopefully the line-by-line comments will help.
One major issue was that you hadn't added the files into the CMakeLists.txt file in the MPAGSCipher directory, so it wasn't even trying to compile it.  So it would have been hard to spot errors.
If you had also implemented some tests then it might have helped to find the problems.

### CODE REVIEW

File: MPAGSCipher/CMakeLists.txt [hunk 1]
```diff
@@ -4,6 +4,8 @@
 add_library(MPAGSCipher STATIC
   CaesarCipher.hpp
   CaesarCipher.cpp
+  PlayfairCipher.hpp
+  PlayfairCipher.cpp
```
Comment (tomlatham) [inline MPAGSCipher/CMakeLists.txt]: Without these lines it won't attempt to compile your new class.

File: MPAGSCipher/PlayfairCipher.cpp [hunk 2]
```diff
@@ -3,27 +3,31 @@
...
     std::transform(key_.begin(),key_.end(),key_.begin(),::toupper);
+
     //Remove non-alpha chars
     auto func=[] (char character){
         if(!isalpha(character))
             return true;
         else
             return false;
-    }
+    };
```
Comment (tomlatham) [inline MPAGSCipher/PlayfairCipher.cpp]: Missing semicolon

File: MPAGSCipher/PlayfairCipher.cpp [hunk 3]
```diff
@@ -36,7 +40,7 @@ void PlayfairCipher::setKey(const std::string& key)
     //Remove duplicated letters
     std::string a{""};
     auto dupletfunc=[&] (char dupletchar){
-        if(a.find(dupletchar)=="npos"){
+        if(a.find(dupletchar)==std::string::npos){
```
Comment (tomlatham) [inline MPAGSCipher/PlayfairCipher.cpp]: std::string::npos is a defined constant that indicates a index that is not within the string

File: MPAGSCipher/PlayfairCipher.cpp [hunk 4]
```diff
@@ -46,108 +50,118 @@ void PlayfairCipher::setKey(const std::string& key)
...
     key_.erase(iterduplet,key_.end());
+
     //store the coords of each letter
-    using Letter2CoordMap=std::map<char, int>;
-    Letter2CoordMap letterstocoord;
-    for(int i{0};i<25;++i){
-        letterstocoord[a[i]]=i;
+    for ( size_t i{0}; i < key_.size(); ++i ) {
+	charLookup_[ key_[i] ] = i;
+	coordLookup_[ i ] = key_[i];
```
Comment (tomlatham) [inline MPAGSCipher/PlayfairCipher.cpp]: Use maps that are defined as private members variables of the class.
I've stuck with your method of having the maps relate the character to the index within the key rather than converting directly to the row and column number at this point.  See our solution for that alternative.

File: MPAGSCipher/PlayfairCipher.cpp [hunk 5]
```diff
@@ -46,108 +50,118 @@ void PlayfairCipher::setKey(const std::string& key)
...
+	    tmpText += outputText[i+1];
+
+	} else {
+	    // Otherwise, if two characters in the digraph are the same,
+	    // we instead add an X (or a Q if the first was an X)
+	    tmpText += (outputText[i] == 'X') ? 'Q' : 'X';
+	    // Need to decrement i since the second character in this digraph
+	    // now becomes the first character in the next one
+	    --i;
+	}
```
Comment (tomlatham) [inline MPAGSCipher/PlayfairCipher.cpp]: Best to separately tackle the two issues of replacing J with I and separating identical letters in digraphs.
Your implementation of the latter had some issues - it would separate any neighbouring identical letters even if they were not in the same digraph.

File: MPAGSCipher/PlayfairCipher.cpp [hunk 6]
```diff
@@ -46,108 +50,118 @@ void PlayfairCipher::setKey(const std::string& key)
...
     }
-    //f the size of input is odd, add a trailing Z
-    if(tempText.length()%2!=0){
-        tempText+="Z";
+    outputText = tmpText;
+
+    // If the size of input is odd, add a trailing Z
+    // (or add an X if the last character is already a Z)
+    if(outputText.length()%2!=0){
+	outputText += (outputText[outputText.size()-1] == 'Z') ? 'X' : 'Z';
```
Comment (tomlatham) [inline MPAGSCipher/PlayfairCipher.cpp]: Adding a Z to a string that already ends in a Z leads to having identical letters in a digraph, so need to change what gets added in that situation.

File: MPAGSCipher/PlayfairCipher.cpp [hunk 7]
```diff
@@ -46,108 +50,118 @@ void PlayfairCipher::setKey(const std::string& key)
...
+	index1 = 5*row1 + column1;
+	index2 = 5*row2 + column2;
+
+	// Find the letters associated with the new indices
+	char char1 = coordLookup_.find(index1)->second;
+	char char2 = coordLookup_.find(index2)->second;
+
+	// Make the replacements
+	(*iter)     = char1;
+	(*(iter+1)) = char2;
```
Comment (tomlatham) [inline MPAGSCipher/PlayfairCipher.cpp]: I could not understand your code I'm afraid.
So I've replaced it with something that is hopefully more readable.
See also our solution for a slightly different approach (due to using the row and column numbers already in the maps).

File: MPAGSCipher/PlayfairCipher.hpp [hunk 8]
```diff
@@ -15,11 +15,41 @@
...
+     * Apply the cipher to the provided text
+     *
+     * \param inputText the text to encrypt or decrypt
+     * \param cipherMode whether to encrypt or decrypt the input text
+     * \return the result of applying the cipher to the input text
+     */
     std::string applyCipher(const std::string& inputText, const CipherMode cipherMode) const;
+
     private:
+    /// The cipher key
```
Comment (tomlatham) [inline MPAGSCipher/PlayfairCipher.hpp]: Include documentation of all member functions and variables.

File: MPAGSCipher/PlayfairCipher.hpp [hunk 9]
```diff
@@ -15,11 +15,41 @@
...
     std::string key_;
 
+    // Lookup tables generated from the key
+
+    /// Lookup table to go from the character to the index in the key
+    std::map<char,size_t> charLookup_;
+
+    /// Lookup table to go from the key index to the character
+    std::map<size_t,char> coordLookup_;
+
```
Comment (tomlatham) [inline MPAGSCipher/PlayfairCipher.hpp]: Include the maps as private member variables.

File: MPAGSCipher/ProcessCommandLine.cpp [hunk 10]
```diff
@@ -73,6 +73,7 @@ bool processCommandLine(const std::vector<std::string>& args,
           else if(args[i+1]=="playfair"){
               settings.cipherType=CipherType::Playfair;
           }
+	  ++i;
```
Comment (tomlatham) [inline MPAGSCipher/ProcessCommandLine.cpp]: This missing increment causes the argument to the --cipher option to be looked at next, instead of whatever follows.

File: README.md [hunk 11]
```diff
@@ -101,15 +109,30 @@ MPAGS-Code
...
-├── mpags-cipher.cpp
+├── Testing               Subdirectory for testing the MPAGSCipher library
+│   ├── CMakeLists.txt
+│   ├── catch.hpp
+│   ├── testCaesarCipher.cpp
+│   ├── testCatch.cpp
+│   ├── testHello.cpp
+│   ├── testPlayfairCipher.cpp
+│   ├── testProcessCommandLine.cpp
+│   └── testTransformChar.cpp
```
Comment (tomlatham) [inline README.md]: Update documentation to match the new state of the code.

File: Testing/testPlayfairCipher.cpp [hunk 12]
```diff
@@ -0,0 +1,15 @@
...
+
+TEST_CASE("Playfair Cipher encryption", "[playfair]") {
+  PlayfairCipher cc{"hello"};
+  REQUIRE( cc.applyCipher("BOBISSOMESORTOFJUNIORCOMPLEXXENOPHONEONEZEROTHING", CipherMode::Encrypt) == "FHIQXLTLKLTLSUFNPQPKETFENIOLVSWLTFIAFTLAKOWATEQOKPPA");
+}
+
+TEST_CASE("Playfair Cipher decryption", "[playfair]") {
+  PlayfairCipher cc{"hello"};
+  REQUIRE( cc.applyCipher("FHIQXLTLKLTLSUFNPQPKETFENIOLVSWLTFIAFTLAKOWATEQOKPPA", CipherMode::Decrypt) == "BOBISXSOMESORTOFIUNIORCOMPLEXQXENOPHONEONEZEROTHINGZ");
+}
```
Comment (tomlatham) [inline Testing/testPlayfairCipher.cpp]: Add some tests.  Ideally these should try and catch all corner cases.  We've just implemented a couple of cases here.

File: Testing/testProcessCommandLine.cpp [hunk 13]
```diff
@@ -27,7 +27,7 @@ TEST_CASE("Version found correctly", "[commandline]") {
 TEST_CASE("All other options processed correctly", "[commandline]") {
 
   ProgramSettings prog_set{false, false, "", "", "", CipherMode::Encrypt, CipherType::Caesar};
-  std::vector<std::string> cmd_line = {"mpags-cipher", "-i", "test-in.txt", "-o", "test-out.txt", "-k", "10", "--decrypt"};
+  std::vector<std::string> cmd_line = {"mpags-cipher", "-i", "test-in.txt", "-o", "test-out.txt", "-k", "10", "--decrypt", "--cipher", "playfair"};
```
Comment (tomlatham) [inline Testing/testProcessCommandLine.cpp]: Without this option, this test fails.
